qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
 
         self._samples = np.append(self._samples, samples)
 
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
 
-def signal_multiply(
-    sig1: Union[BaseSignal, float, int, complex], sig2: Union[BaseSignal, float, int, complex]
-) -> Signal:
-    r"""Implements mathematical multiplication between two signals.
-    Since a signal is represented by
-
-    .. math::
-
-        \Omega(t)*exp(2\pi i \nu t)
+        return "DiscreteSignal(dt={dt}, carrier_freq={freq}, phase={phase})".format(
+            dt=self.dt, freq=str(self.carrier_freq), phase=str(self.phase)
+        )
 
-    multiplication of two signals implements
 
-    .. math::
+class SignalSum(Signal):
+    """Class representing a sum of ``Signal`` objects."""
 
-        \Omega_1(t)*\Omega_2(t)*exp(2\pi i (\nu_1+\nu_2) t)
+    def __init__(self, *args, name: Optional[str] = None):
+        """Initialize with a list of Signal objects through ``args``.
 
+        Args:
+            args: ``Signal`` subclass objects.
+            name: Name of the sum.
 
-    Args:
-        sig1: A child of base signal or a constant.
-        sig2: A child of base signal or a constant.
+        Raises:
+            QiskitError: If ``args`` are not subclasses of ``Signal``.
+        """
+        self._name = name
 
-    Returns:
-        signal: The type will depend on the given base class.
-    """
+        self.components = []
+        for sig in args:
+            if isinstance(sig, SignalSum):
+                self.components += sig.components
+            elif isinstance(sig, Signal):
+                self.components.append(sig)
+            else:
+                raise QiskitError(
+                    "Components of a SignalSum must be instances of a Signal subclass."
+                )
+
+        self._envelopes = [sig.envelope for sig in self.components]
+
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
+
+        carrier_freqs = []
+        for sig in self.components:
+            carrier_freqs.append(sig.carrier_freq)
+
+        phases = []
+        for sig in self.components:
+            phases.append(sig.phase)
+
+        # set carrier and phase
+        self.carrier_freq = carrier_freqs
+        self.phase = phases
+
+    def envelope(self, t: Union[float, np.array, Array]) -> Array:
+        """Evaluate envelopes of each component. For vectorized operation,
+        last axis indexes the envelope, and all proceeding axes are the
+        same as the ``t`` arg.
+        """
```

Comment:
This a bit unclear.

Reply:
I've updated this. I've actually removed most of the description - what I was describing was basically what "vectorized" means, which maybe isn't actually necessary here.